### DISCUSSION TITLE
Bug or feature? Can't change port for MQTT with TLS.

### DISCUSSION THREAD

**Discussion by Kugeleis** [arendst/Tasmota#18151] [Q&A]
I compiled tasmota 12.4 with [TLS for MQTT](https://tasmota.github.io/docs/TLS/). In the device GUI I can't change the default MQTT port from 8883 to something else. Whatever I type in is lost after 'save'. Is this a wanted behavior?

**Comment by s-hadinger** [comment]
No it's not. Can you try with `MqttPort` command in the console.

**Comment by Kugeleis** [comment]
I typed `mqttport 1883` on the console and got this:
```
00:00:00.001 HDW: ESP8266EX
00:00:00.050 CFG: aus Flash geladen von FA, Anzahl 2760
00:00:00.055 QPC: Count 1
00:00:00.083 Projekt tasmota - Version 12.4.0(TasmoCompiler-esp8266generic)-2_7_4_9(2023.03.08 18:32:11)
00:00:00.534 WIF: verbinden mit AP2 xxxx Channel 11 BSSId EC:A9 in Modus 11n als cfre-5807...
00:00:01.752 WIF: verbunden
00:00:02.005 HTP: Web-Server aktiv bei cfre-5807 mit IP-Adresse 192.168.1.181
00:00:02.410 CMD: mqttport
00:00:02.415 RSL: RESULT = {"MqttPort":8883}
14:45:04.027 MQT: Verbindungsversuch...
14:45:04.029 MQT: Authenticating with an Azure IoT Hub Token
14:45:04.042 MQT: Verbindung fehlgeschlagen aufgrund von 192.168.1.88:8883, rc -2. Wiederversuch in 10 s
14:45:07.457 QPC: Reset
```
It automatically assumes I try to access the azure IOT and resets the port to 8883. In the GUI "MQTT TLS" is not checked. My compile options are
```
#ifndef USE_MQTT_TLS
#define USE_MQTT_TLS
#endif
#define USE_MQTT_AZURE_IOT
```

**Reply by s-hadinger** [reply]
Ah. The person who developed Azure iot integration hard-coded the port number.

**Comment by Kugeleis** [comment]
Where can I address this issue?

**Reply by barbudor** [reply]
I would assume that it is hard coded because you can't change it on the Azure side
From Microsoft docs it is specified that MQTT is listenning on port 8883
https://learn.microsoft.com/en-gb/azure/iot-hub/iot-hub-devguide-protocols#port-numbers

Are you able to change the MQTT port on Azure side ?

https://github.com/arendst/Tasmota/blob/f87cd7859a193595fc52f56e2b5aa2676e3c8ccc/tasmota/tasmota_xdrv_driver/xdrv_02_9_mqtt.ino#L32-L34

**Reply by arendst** [reply]
It's even worse. The Azure person also forced the settings->mqtt_port to 8883:
```
#ifdef USE_MQTT_AZURE_IOT
  Settings->mqtt_port = 8883;
#endif //USE_MQTT_AZURE_IOT
```
So the question here is if it's even possible to have a different Azure MQTT port on the server side. If not, then there is no issue.

**Reply by sfromis** [reply]
It is not clear if you actually want to use Azure or not. If you *do* want that, it does not look like it makes sense to try with a different port number.

If you just want to use MQTT-TLS without the broker being Azure, you should not have `#define USE_MQTT_AZURE_IOT` enabled in your build.

**Comment by Kugeleis** [comment]
The problem is that this compile option breaks an existing local connection via port 1883. I'd expect I still can use the device with a local mqtt broker. But this hard coded 8883 breaks my setup. No other option forces me to use it and renders an input field useless.

**Reply by barbudor** [reply]
This is the point : if you want to use a local broker you should NOT use #define USE_MQTT_AZURE_IOT

You can't connect to 2 brokers at the same time
Beside, enabling the Azure IOT mode makes a lot of changes in the way Tasmota is handling MQTT
This is why these are 2 completely exclusive working modes

**Reply by sfromis** [reply]
If you need to connect to Azure MQTT broker on another device, you'll need a different binary for that device, not the same as for other brokers like local.
